ClosedXML/Excel/Ranges/Index/XLRangeIndex.cs

```diff
@@ -0,0 +1,484 @@
+                    {
+                        res |= childQuadrant.Add(range);
+                        addToChild = true;
+                        break;
+                    }
+                }
+
+                if (!addToChild)
+                    res = AddInternal(range);
+
+                if (ChildQuadrants == null && addToChild)
+                    ChildQuadrants = childQuadrants;
+
+                return res;
+            }
+
+            /// <summary>
+            /// Get all ranges from the quadrant and all child quadrants (recursively).
+            /// </summary>
+            public IEnumerable<IXLRangeBase> GetAll()
+            {
+                if (Ranges != null)
+                {
+                    foreach (var range in Ranges)
+                        yield return range;
+                }
+
+                if (ChildQuadrants != null)
+                {
+                    foreach (var childQuadrant in ChildQuadrants)
+                    {
+                        var childRanges = childQuadrant.GetAll();
+                        foreach (var range in childRanges)
+                            yield return range;
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Get all ranges from the quadrant and all child quadrants (recursively) that intersect the specified address.
+            /// </summary>
+            public IEnumerable<IXLRangeBase> GetIntersectedRanges(IXLRangeAddress rangeAddress)
+            {
+                if (Ranges != null)
+                {
+                    foreach (var range in Ranges)
+                    {
+                        if (range.RangeAddress.Intersects(rangeAddress))
+                            yield return range;
+                    }
+                }
+
+                if (ChildQuadrants != null)
+                {
+                    foreach (var childQuadrant in ChildQuadrants)
+                    {
+                        if (childQuadrant.Intersects(in rangeAddress))
+                        {
+                            var childRanges = childQuadrant.GetIntersectedRanges(rangeAddress);
+                            foreach (var range in childRanges)
+                                yield return range;
+                        }
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Get all ranges from the quadrant and all child quadrants (recursively) that cover the specified address.
+            /// </summary>
+            public IEnumerable<IXLRangeBase> GetIntersectedRanges(IXLAddress address)
+            {
+                if (Ranges != null)
+                {
+                    foreach (var range in Ranges)
+                    {
+                        if (range.RangeAddress.Contains(address))
+                            yield return range;
+                    }
+                }
+
+                if (ChildQuadrants != null)
```

Comment:
I do not pre-initialize `ChildQuadrants` and `Ranges` on purpose. They are initialized only when something is added to the particular quadrant; otherwise, we'd end up having a tree with ~128 million nodes, most of them empty.

Reply:
Ok, that's fair.